ras/saa/tsang.py

```diff
@@ -0,0 +1,151 @@
+import time
+from itertools import product
+
+import gurobipy as gp
+from gurobipy import GRB
+from numpy.random import default_rng
+
+from ras.appointment.true_optimal import compute_optimal_schedule
+from ras.classes import CostEvaluator, ProblemData, Solution
+from ras.Result import Result
+
+from .utils import _sample_distance_matrices, _sample_service_times
+
+
+def tsang(
+    seed: int,
+    data: ProblemData,
+    cost_evaluator: CostEvaluator,
+    max_runtime: int,
+    num_scenarios: int,
+    **kwargs,
+) -> Result:
+    """
+    Solves the routing and appointment scheduling problem using the sample
+    average approximation method from Tsang and Shehadeh (2022).
+
+    Parameters
+    ----------
+    seed
+        The random seed to use for the solver (unused).
+    data
+        The problem data.
+    cost_evaluator
+        The cost evaluator to use for computing the cost of a schedule.
+    max_runtime
+        The maximum runtime.
+    num_scenarios
+        The number of scenarios to use for the sample average approximation.
+
+    Returns
+    -------
+    Result
+        The algorithm results.
+    """
+    start = time.perf_counter()
+
+    m = gp.Model()
+    N = data.dimension
+    S = num_scenarios
+
+    rng = default_rng(seed)
+
+    distance = _sample_distance_matrices(data, num_scenarios, rng)
+
+    if num_scenarios == 1:
+        distance[:, :, 0] = data.distances
+
+    service = _sample_service_times(data, num_scenarios, rng)
+
+    x = m.addVars(N, N, vtype=GRB.BINARY, name="order")
+    a = m.addVars(N, vtype=GRB.CONTINUOUS, lb=0, name="appointment time")
+    idle = m.addVars(N, S, vtype=GRB.CONTINUOUS, lb=0, name="idle")
+    wait = m.addVars(N, S, vtype=GRB.CONTINUOUS, lb=0, name="wait")
+    total_travel = m.addVars(S, vtype=GRB.CONTINUOUS, lb=0, name="total_travel")
+
+    # Objective
+    travel_weight = cost_evaluator.travel_weight
+    idle_weight = cost_evaluator.idle_weight
+    wait_weights = cost_evaluator.wait_weights
+
+    avg_travel = sum(travel_weight * total_travel[s] for s in range(S)) / S
+    avg_idle = sum(idle_weight * idle[i, s] for i in range(N) for s in range(S)) / S
+
+    # TODO this is wrong in the paper
+    avg_wait = sum(wait_weights[i] * wait[i, s] for i in range(N) for s in range(S)) / S
```

Comment:
I'm pretty sure this is wrong in the paper. The variable $w_j$ is the j-th served customer of the solution. The weight $c_j^W$ refers to the weight of customer $j$. This expression is therefore incorrect since it assigns client weights to possibly other clients.

![image](https://github.com/leonlan/routing-appointment-scheduling/assets/30997278/1b457c74-c99f-4c2b-a97e-ccba18d07641)

Reply:
Adding this can work with customer-specific weights, but it makes the model terribly slow.

```python
wait_client = m.addVars(N, S, vtype=GRB.CONTINUOUS, lb=0, name="wait_client")

avg_wait = (
    sum(wait_weights[i] * wait_client[i, s] for i in range(N) for s in range(S)) / S
)

# Waiting time of client i is not the same as waiting time at location j.
for i in range(1, N):
    for s in range(S):
        m.addConstr(
            wait_client[i, s] == sum(wait[j, s] * x[i, j] for j in range(1, N))
        )
```